File: articles/restoration-ecology-repair-of-living-systems/go/restoration_cli.go
package main

import (
	"fmt"
	"math"
)

type Scenario struct {
	Name        string
	S           float64
	B           float64
	Disturbance float64
}

type Result struct {
	FinalV float64
	FinalM float64
	FinalF float64
	PeakF  float64
}

func simulateRestoration(s float64, bSupport float64, disturbance float64) Result {
	a := 0.8
	b := 0.15
	c := 0.20
	p := 0.10
	q := 0.25
	r := 0.12
	u := 0.08
	v := 0.10
	w := 0.18
	dt := 0.05
	tEnd := 50.0

	vegetation := 10.0
	microbial := 8.0
	function := 6.0
	peakFunction := function

	for time := dt; time <= tEnd+1e-12; time += dt {
		dv := a*s - b*vegetation - c*disturbance
		dm := p*vegetation + q*bSupport - r*microbial
		df := u*vegetation + v*microbial - w*disturbance

		vegetation = math.Max(0.0, vegetation+dv*dt)
		microbial = math.Max(0.0, microbial+dm*dt)
		function = math.Max(0.0, function+df*dt)

		peakFunction = math.Max(peakFunction, function)
	}

	return Result{
		FinalV: vegetation,
		FinalM: microbial,
		FinalF: function,
		PeakF:  peakFunction,
	}
}

func main() {
	scenarios := []Scenario{
		{"low_effort_high_disturbance", 0.7, 0.8, 0.8},
		{"moderate_effort_moderate_disturbance", 1.0, 0.8, 0.5},
		{"high_effort_low_disturbance", 1.4, 0.8, 0.2},
		{"soil_limited_recovery", 1.1, 0.3, 0.4},
	}

	fmt.Println("scenario,final_V,final_M,final_F,peak_F")

	for _, scenario := range scenarios {
		result := simulateRestoration(scenario.S, scenario.B, scenario.Disturbance)
		fmt.Printf("%s,%.6f,%.6f,%.6f,%.6f\n",
			scenario.Name,
			result.FinalV,
			result.FinalM,
			result.FinalF,
			result.PeakF,
		)
	}
}

File: articles/restoration-ecology-repair-of-living-systems/rust/restoration_cli.rs
// Restoration CLI in Rust

fn simulate_restoration(s: f64, b_support: f64, disturbance: f64) -> (f64, f64, f64, f64) {
    let a = 0.8;
    let b = 0.15;
    let c = 0.20;
    let p = 0.10;
    let q = 0.25;
    let r = 0.12;
    let u = 0.08;
    let v = 0.10;
    let w = 0.18;
    let dt = 0.05;
    let t_end = 50.0;

    let mut vegetation = 10.0;
    let mut microbial = 8.0;
    let mut function = 6.0;
    let mut peak_function = function;

    let mut time = dt;

    while time <= t_end + 1e-12 {
        let dv = a * s - b * vegetation - c * disturbance;
        let dm = p * vegetation + q * b_support - r * microbial;
        let df = u * vegetation + v * microbial - w * disturbance;

        vegetation = f64::max(0.0, vegetation + dv * dt);
        microbial = f64::max(0.0, microbial + dm * dt);
        function = f64::max(0.0, function + df * dt);

        peak_function = f64::max(peak_function, function);
        time += dt;
    }

    (vegetation, microbial, function, peak_function)
}

fn main() {
    let scenarios = [
        ("low_effort_high_disturbance", 0.7, 0.8, 0.8),
        ("moderate_effort_moderate_disturbance", 1.0, 0.8, 0.5),
        ("high_effort_low_disturbance", 1.4, 0.8, 0.2),
        ("soil_limited_recovery", 1.1, 0.3, 0.4),
    ];

    println!("scenario,final_V,final_M,final_F,peak_F");

    for (name, s, b_support, disturbance) in scenarios {
        let (v, m, f, peak_f) = simulate_restoration(s, b_support, disturbance);
        println!("{},{:.6},{:.6},{:.6},{:.6}", name, v, m, f, peak_f);
    }
}
